Automattic/sensei
# Add restricted content blocks

### Changes proposed in this Pull Request

* This PR introduces a new blocks to the single course page, 'Restrict Content' (naming suggestions welcome 😄 ).
* The block is similar to how the 'Group' block works but they show the content to enrolled, unenrolled or users which completed a course only.
* I investigated to possibly add the blocks to the single lesson page too. Currently in the single lesson template we don't show the content at all when the user is not enrolled. For this reason we should probably wait until we drop the single lesson template to see if it fits to add these blocks.

_Update:_
* It is possible to transform any selection of blocks to move it inside a wrapper 'Enrolled Content' or 'Unenrolled Content' block. This allows users easily to restrict the content after it is created. When transforming one of these blocks to the inverse one, a wrapper is not created and the block is switched instead.
![record](https://user-images.githubusercontent.com/53191348/102106933-d571e000-3e39-11eb-8fc1-d8fc40a94e43.gif)

_Second Update:_
* There is now a single block for all cases. The user can select between different options by using a dropdown. It also has a new option which restricts the content to users which completed a course.
![button](https://user-images.githubusercontent.com/53191348/104202081-22de6e80-5433-11eb-93fa-2ffc09b607ef.gif)
* Since there wasn't an intuitive option to remove the restriction once the block is created, I also added an option to the drop down menu which removes the 'Restrict content' block. Basically it replaces the block whith its inner blocks
![remove](https://user-images.githubusercontent.com/53191348/104202300-6a64fa80-5433-11eb-87e9-dfa6b9fc26e3.gif)
* I have considered the suggestions in [this comment](https://github.com/Automattic/sensei/pull/3849#issuecomment-748191152) but I decided against them for the following reasons:
  * The toolbar should follow gutenberg's minimalistic approach so any not absolutely necessary text should be avoided. The ideal way to do this would be to have the selected icon on the toolbar and the options on the dropdown should be a button with the icon and a few words, similar with the rest of the dropdown menus. Additionally this would require more time to implement.
  * I used text only for now to avoid the extra work of creating new icons.
  * Since the button can be considered as a title too, I think that we don't need any extra indicator to explain that the content will be hidden. We can revise this if users find the block confusing.


### Testing instructions

* In a course include the new block and add some content in them.
* Try the 'Enrolled Users', 'Unenrolled Users' and 'Course Completed' options and observe that the content is displayed (or not) accordingly in the frontend.
* Try selecting multiple blocks and transform them to a 'Restrict content' block.
* Try removing the restriction from a block.
